src/display/display_utils.js

```diff
@@ -274,8 +280,163 @@ class DOMFilterFactory extends BaseFilterFactory {
+
+    if (!fgColor || !bgColor) {
+      return this.#hcmHighlightUrl;
+    }
+
+    let fgRGB, bgRGB, newFgRGB, newBgRGB;
+    const colors = [
+      fgColor,
+      null,
+      bgColor,
+      null,
+      newFgColor,
+      null,
+      newBgColor,
+      null,
+    ];
+    for (let i = 0; i < 8; i += 2) {
+      let color = colors[i];
+      this.#defs.style.color = color;
+      color = getComputedStyle(this.#defs).getPropertyValue("color");
+      const rgb = (colors[i + 1] = getRGB(color));
+      colors[i] = Util.makeHexColor(...rgb);
+    }
+    this.#defs.style.color = "";
+    [
+      fgColor,
+      fgRGB,
+      bgColor,
+      bgRGB,
+      newFgColor,
+      newFgRGB,
+      newBgColor,
+      newBgRGB,
+    ] = colors;
+
+    // Now we can create the filters to highlight some canvas parts.
+    // The colors in the pdf will almost be Canvas and CanvasText, hence we
+    // want to filter them to finally get Highlight and HighlightText.
+    // Since we're in HCM the background color and the foreground color should
+    // be really different when converted to grayscale (if they're not then it
+    // means that we've a poor contrast). Once the canvas colors are converted
+    // to grayscale we can easily map them on their new colors.
+    // The grayscale step is important because if we've something like:
+    //   fgColor = #FF....
+    //   bgColor = #FF....
+    //   then we are enable to map the red component on the new red components
+    //   which can be different.
+
+    const fgGray = 0.2126 * fgRGB[0] + 0.7152 * fgRGB[1] + 0.0722 * fgRGB[2];
+    const bgGray = 0.2126 * bgRGB[0] + 0.7152 * bgRGB[1] + 0.0722 * bgRGB[2];
+
+    let getSteps;
+    if (fgGray < bgGray) {
+      getSteps = (fg, bg, n) => {
+        const arr = new Array(256);
+        const step = (bgGray - fgGray) / n;
+        const newStart = fg / 255;
+        const newStep = (bg - fg) / (255 * n);
+        let prev = 0;
+        for (let i = 0; i <= n; i++) {
+          const k = Math.round(fgGray + i * step);
+          const value = newStart + i * newStep;
+          for (let j = prev; j <= k; j++) {
+            arr[j] = value;
+          }
+          prev = k + 1;
+        }
+        for (let i = prev; i < 256; i++) {
+          arr[i] = arr[prev - 1];
+        }
+        return arr.join(",");
+      };
+    } else {
+      getSteps = (fg, bg, n) => {
+        const arr = new Array(256);
+        const step = (bgGray - fgGray) / n;
+        const newStart = fg / 255;
+        const newStep = (bg - fg) / (255 * n);
+        let prev = 256;
+        for (let i = 0; i <= n; i++) {
+          const k = Math.round(fgGray + i * step);
+          const value = newStart + i * newStep;
+          for (let j = k; j < prev; j++) {
+            arr[j] = value;
+          }
+          prev = k;
+        }
+        for (let i = 0; i < prev; i++) {
+          arr[i] = arr[prev];
+        }
+        return arr.join(",");
+      };
+    }
```

Comment:
Could we please reduce some duplication here, since considering how little this code is being used it seems reasonable to focus on reducing the code-size. (The `addHighlightHCMFilter`-method will only be called *at most once* per PDF document, and not at all for most users.)

```suggestion
    const getSteps = (fg, bg, n) => {
      const arr = new Array(256);
      const step = (bgGray - fgGray) / n;
      const newStart = fg / 255;
      const newStep = (bg - fg) / (255 * n);
      if (fgGray < bgGray) {
        let prev = 0;
        for (let i = 0; i <= n; i++) {
          const k = Math.round(fgGray + i * step);
          const value = newStart + i * newStep;
          for (let j = prev; j <= k; j++) {
            arr[j] = value;
          }
          prev = k + 1;
        }
        for (let i = prev; i < 256; i++) {
          arr[i] = arr[prev - 1];
        }
      } else {
        let prev = 256;
        for (let i = 0; i <= n; i++) {
          const k = Math.round(fgGray + i * step);
          const value = newStart + i * newStep;
          for (let j = k; j < prev; j++) {
            arr[j] = value;
          }
          prev = k;
        }
        for (let i = 0; i < prev; i++) {
          arr[i] = arr[prev];
        }
      }
      return arr.join(",");
    };
```